[쪼밀리] 로또 2단계 - 수동 미션 제출합니다

chomily opened the pull request:
2단계 수동 미션 제출합니다!
감사합니다 :)

chomily committed:
refactor: DTO 클래스 내부 부적절한 상수 삭제 및 View 관련 메서드 이동

chomily committed:
docs: README 구현 기능 목록 업데이트

- fix: 이전 커밋에 의한 리턴 타입 오류 수정

chomily committed:
feat: 입력 번호로 로또 생성 기능 추가 및 LottoMachine 클래스 통합

- AutoLottMachine, LottoMachineForTest 클래스 제거
- 불필요한 DTO 테스트 제거
- 불필요한 테스트 제거

chomily committed:
feat: 수동으로 구매할 로또 수, 로또 번호 입력 기능

- 수동구매용 createOneTicket() 매개변수 타입 변경

chomily committed:
feat: LottoMachine 분리 -> Money 클래스 작성

- 투입 금액 유효성 검사(1000원 이상) 책임 이동
- 투입 금액을 초과하는 수동 로또 구매 희망시 예외 처리
- 테스트 코드 추가

chomily committed:
feat: 구매가능한 티켓 수 계산, 자동 로또 지출 기능

- 테스트 코드 추가

chomily committed:
feat: 수동 로또 구매 기능 추가, 컨트롤러, 서비스 반영

- NumberOfTicketDTO, ManualNumbersDTO, PurchaseStatusDTO, StatisticsDTO 작성

chomily committed:
refactor: LottoMachine 추상화 및 구현체 AutoLottoMachine, ManualLottoMachine 작성

kang-hyungu commented on src/main/java/lotto/view/dto/ManualNumbersDTO.java:
```diff
@@ -0,0 +1,15 @@
+import java.util.List;
+
+public class ManualNumbersDTO {
+    private List<List<Integer>> lists;
```
lists가 뭘까요?
좀더 의미 있는 변수명을 써주세요

kang-hyungu commented on src/main/java/lotto/controller/LottoController.java:
```diff
@@ -1,38 +1,65 @@
+
+    private LottoTicketBundle createLottoTicketBundle(Money money) {
+        List<LottoTicket> manualTickets = createManualTickets(money);
+        List<LottoTicket> autoTickets = createAutoTickets(money);
```
autoTickets 생성과 manualTickets 생성 순서를 바꾸면 에러가 발생하네요
autoTickets 생성이 manualTickets에 의존하지 않도록 수정해보아요

kang-hyungu commented on src/main/java/lotto/controller/LottoController.java:
```diff
@@ -1,38 +1,65 @@
+        return new Money(bettingMoney.getBettingMoney());
+    }
+
+    private LottoTicketBundle createLottoTicketBundle(Money money) {
```
LottoController 클래스에 아래 static 변수를 추가하고
```java
private static final List<LottoMachine> lottoMachines = Arrays.asList(new AutoLottoMachine(), new ManualLottoMachine());
```

새로운 LottoMachine이 추가되도 이 메서드를 수정하지 않도록 개선해볼까요?
LottoService에서 수동, 자동 각각 생성하는 코드가 중복이니 이 부분을 제거할 방법을 찾아보면 개선 가능할거에요

kang-hyungu commented on src/main/java/lotto/service/LottoService.java:
```diff
@@ -1,41 +1,52 @@
 
-        return lottoMachine.createWinLottoTicket(winNumbers, bonusNumber);
+    public WinLottoTicket createWinLottoTicket(List<Integer> winNumbers, int bonusNumber) {
+        return defaultAutoMachine.createWinLottoTicket(winNumbers, bonusNumber);
```
AutoMachine에서 당첨 로또를 뽑는건가요?
수동 머신에서는 당첨자가 안나오는 걸까요?

chomily replied:
LottoMachine에서 WinLottoTicket생성하는 기능을 제거하고 생성 책임을 WinLottoTicket에게 온전히 주었습니다. 그리고 service단에서 직접 당첨 로또를 생성하도록 변경하였습니다.

kang-hyungu reviewed (changes requested):
피드백 반영 잘 하셨어요
이번엔 피드백 많이 남기진 않았는데, 대신 중복을 제거할 수 있는 방법을 제안해봤어요
한 번 고민해서 수정해보시고 재요청 주세요
수정하면서 궁금한 점 생기면 코멘트나 dm 남겨주세요

chomily committed:
refactor: Money 클래스 구조 수정으로 로또 수동/자동 생성 순서 의존성 분리

chomily committed:
refactor: WinLottoTicket 생성자 구조 변경 및 LottoMachine의 책임 분리

chomily committed:
refactor: LottoMachine 인터페이스화, Customer 클래스 작성

- 수동 로또 번호 전달 방식 변경으로 중복 코드 제거 및 확장성 확보
  (ManualLottoMachine 생성자 -> Customer 클래스 인스턴스 변수)

chomily committed:
refactor: Service, Controller 구조 변경 / RequestDTO 생성 위치 변경

chomily committed:
refactor: 네이밍 변경 Money -> PurchaseInfo

kang-hyungu commented on src/main/java/lotto/domain/ticket/AutoLottoMachine.java:
```diff
@@ -1,19 +1,32 @@
+        NullOrEmptyValidator.isNull(customer);
+
+        List<LottoTicket> lottoTickets = new ArrayList<>();
+        for (int i = 0; i < customer.getPurchaseInfo().getNumberOfLeftTickets(); i++) {
```
객체지향 생활체조
규칙 4: 한 줄에 점을 하나만 찍는다.

kang-hyungu commented on src/main/java/lotto/domain/ticket/ManualLottoMachine.java:
```diff
@@ -0,0 +1,42 @@
+        int numberOfManualTickets = customer.getPurchaseInfo().getNumberOfManualTickets();
+        List<List<Integer>> manualNumbers = customer.getManualNumbers();
+
+        NullOrEmptyValidator.isNullOrEmpty(manualNumbers);
```
customer 객체 생성자에서 manualNumbers가 null이 아닌지 체크를 하는게 좋겠네요
안그러면 지금처럼 getManualNumbers 할 때마다 null 여부를 체크해야되니깐요
객체 생성할 때 검증이 끝내도록 설계하면 이 객체가 반환하는 값은 검증할 필요 없이 믿고 쓸 수 있는 데이터가 될거에요

kang-hyungu commented on src/main/java/lotto/domain/ticket/ManualLottoMachine.java:
```diff
@@ -0,0 +1,42 @@
+    @Override
+    public List<LottoTicket> buyTickets(Customer customer) {
+        NullOrEmptyValidator.isNull(customer);
+        int numberOfManualTickets = customer.getPurchaseInfo().getNumberOfManualTickets();
```
규칙 4: 한 줄에 점을 하나만 찍는다.

kang-hyungu commented on src/main/java/lotto/domain/customer/Customer.java:
```diff
@@ -0,0 +1,21 @@
+    private final PurchaseInfo purchaseInfo;
+    private final List<List<Integer>> manualNumbers;
+
+    public Customer(PurchaseInfo purchaseInfo, List<List<Integer>> manualNumbers) {
```
이 파라미터들도 유효한 값인지 검사가 필요하겠네요

kang-hyungu reviewed (changes requested):
피드백 반영 잘 하셨어요 💯 
로또 번호 캐시, 객체 조합, 의존 주입까지 잘 활용했어요 👍 
아주 간단한 피드백 몇 개만 남겼어요
수정해서 재요청 주시면 빠르게 머지할게요

chomily committed:
fix/refactor: 수동 로또 0장 요구 시 버그 픽스 및 유효성 검사 보강

- manualNumbers 유효성 검사 위치 변경 ManualLottoMachine -> Customer
- (fix)수동 로또 구매 0장 입력 시, 수동 로또 번호 입력 요구 Skip
- Customer 생성자 추가
- 테스트 추가

kang-hyungu reviewed (approved):
피드백 반영 잘 하셨어요 👍 
이번 미션은 여기서 마무리 지어도 되겠네요
코드 구현하느라 수고하셨어요!
